### DISCUSSION TITLE
How can I forward a request, basically unaltered, to a new URL?

### DISCUSSION THREAD

**Discussion by thomasmost** [rwf2/Rocket#2304] [Questions]
My use-case is that I am writing a public-facing API in Rocket, and would like to forward a Stripe webhook event un-altered to a separate, internal billing-service.

Is there a simple way of passing a request to a new URL, practically unaltered but for an additional header, perhaps?

**Comment by blackghost1987** [comment]
You cannot call another server directly from Rocket, as Rocket is only an HTTP server, and does not include a HTTP client. (well technically it does, but that's just internal stuff for testing). You have to use another crate which implements a HTTP client (such as Reqwest) to make HTTP calls.

It's not a trivial task, there are a couple things to consider, for example:
- Reqwest has a blocking and an async variant, you want to use the right one depending on your use-case and your Rocket version
- also you have to decide if you want to store the request body / response body in memory or if you want to stream the data

There are a bunch of discussions / issues about this, but AFAIK none of them has a fully fledged solution. I hope it gets you started though:
https://github.com/SergioBenitez/Rocket/discussions/2068
https://github.com/SergioBenitez/Rocket/discussions/1983
https://github.com/SergioBenitez/Rocket/issues/1366
https://github.com/SergioBenitez/Rocket/issues/1521

Depending on your use case, it might be an easier solution to "divert" the request before reaching Rocket. You can do this by using a reverse proxy as a previous layer before Rocket, e.g. Nginx or Apache. You should be able to set a special URL path in the proxy config, specifying which requests should go to the internal billing-service, while every other request will go to your Rocket server. These proxies should be able to add additional headers as well, if needed.
This solution has the advantage of actually separating the responsibilities of the different services, also relieve your Rocket server of the extra performance need caused by forwarding the requests.

**Reply by edgarogh** [reply]
Yes, a reverse proxy is the way to go. I also recommend [Caddy](https://caddyserver.com/) which is incredibly easy to use and gives you HTTPS automatically.

The config could look like this
```caddy
example.com

route /api {
    reverse_proxy 127.0.0.1:9000 # your Rocket server
}

route /billing {
    header Custom-Header "My value"
    reverse_proxy billing-service.com
}
```

**Comment by thomasmost** [comment]
Thank you!
